packages/fxa-auth-server/lib/routes/totp.js

```diff
@@ -271,141 +225,110 @@ module.exports = (log, db, mailer, customs, config) => {
+        const sharedSecret = token.sharedSecret;
+        const tokenVerified = token.verified;
+
+        const authenticator = new otplib.authenticator.Authenticator();
+        authenticator.options = Object.assign(
+          {},
+          otplib.authenticator.options,
+          { secret: sharedSecret }
+        );
+        const isValidCode = authenticator.check(code, sharedSecret);
+
+        // Once a valid TOTP code has been detected, the token becomes verified
+        // and enabled for the user.
+        if (isValidCode && !tokenVerified) {
+          await db.updateTotpToken(sessionToken.uid, {
+            verified: true,
+            enabled: true,
           });
 
-        function getTotpToken() {
-          return db.totpToken(sessionToken.uid).then(token => {
-            sharedSecret = token.sharedSecret;
-            tokenVerified = token.verified;
+          await log.notifyAttachedServices('profileDataChanged', request, {
+            uid: sessionToken.uid,
           });
         }
 
-        function verifyTotpCode() {
-          const authenticator = new otplib.authenticator.Authenticator();
-          authenticator.options = Object.assign(
-            {},
-            otplib.authenticator.options,
-            { secret: sharedSecret }
+        // If this is a new registration, replace and generate recovery codes
+        if (isValidCode && !tokenVerified) {
+          recoveryCodes = await db.replaceRecoveryCodes(
+            uid,
+            RECOVERY_CODE_COUNT
           );
-          isValidCode = authenticator.check(code, sharedSecret);
         }
 
-        // Once a valid TOTP code has been detected, the token becomes verified
-        // and enabled for the user.
-        function verifyTotpToken() {
-          if (isValidCode && !tokenVerified) {
-            return db
-              .updateTotpToken(sessionToken.uid, {
-                verified: true,
-                enabled: true,
-              })
-              .then(() => {
-                return log.notifyAttachedServices(
-                  'profileDataChanged',
-                  request,
-                  {
-                    uid: sessionToken.uid,
-                  }
-                );
-              });
-          }
+        // If a valid code was sent, this verifies the session using the `totp-2fa` method.
+        if (isValidCode && sessionToken.authenticatorAssuranceLevel <= 1) {
+          await db.verifyTokensWithMethod(sessionToken.id, 'totp-2fa');
         }
 
-        // If this is a new registration, replace and generate recovery codes
-        function replaceRecoveryCodes() {
-          if (isValidCode && !tokenVerified) {
-            return db
-              .replaceRecoveryCodes(uid, RECOVERY_CODE_COUNT)
-              .then(result => (recoveryCodes = result));
-          }
+        if (isValidCode) {
+          log.info('totp.verified', { uid });
+          request.emitMetricsEvent('totpToken.verified', { uid });
```

Comment:
Should we await on these or not? I think in another file they were waited on.

Reply:
Yes, seems like it is universally awaited elsewhere :+1: